core/src/main/java/org/mule/runtime/core/execution/ExtensionFlowProcessingPhase.java

```diff
@@ -0,0 +1,209 @@
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.execution;
+
+import static org.mule.runtime.core.DefaultEventContext.create;
+import static org.mule.runtime.core.context.notification.ConnectorMessageNotification.MESSAGE_ERROR_RESPONSE;
+import static org.mule.runtime.core.context.notification.ConnectorMessageNotification.MESSAGE_RECEIVED;
+import static org.mule.runtime.core.context.notification.ConnectorMessageNotification.MESSAGE_RESPONSE;
+import static org.mule.runtime.core.execution.TransactionalErrorHandlingExecutionTemplate.createMainExecutionTemplate;
+import org.mule.runtime.api.exception.MuleRuntimeException;
+import org.mule.runtime.api.message.Message;
+import org.mule.runtime.core.api.Event;
+import org.mule.runtime.core.api.exception.MessagingExceptionHandler;
+import org.mule.runtime.core.api.message.InternalMessage;
+import org.mule.runtime.core.api.policy.SourcePolicyParametersTransformer;
+import org.mule.runtime.core.api.source.MessageSource;
+import org.mule.runtime.core.exception.MessagingException;
+import org.mule.runtime.core.policy.NextOperation;
+import org.mule.runtime.core.policy.PolicyProvider;
+import org.mule.runtime.core.policy.OperationPolicy;
+import org.mule.runtime.core.policy.PolicyManager;
+import org.mule.runtime.core.policy.SourcePolicy;
+import org.mule.runtime.core.transaction.MuleTransactionConfig;
+import org.mule.runtime.dsl.api.component.ComponentIdentifier;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import javax.resource.spi.work.Work;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This phase routes the message through the flow.
+ * <p>
+ * To participate of this phase, {@link MessageProcessTemplate} must implement {@link ExtensionFlowProcessingPhaseTemplate}
+ *
+ * This implementation will know how to process messages from extension's sources
+ */
+public class ExtensionFlowProcessingPhase
+    extends NotificationFiringProcessingPhase<ExtensionFlowProcessingPhaseTemplate> {
+
+  private final PolicyManager policyManager;
+  protected transient Logger logger = LoggerFactory.getLogger(getClass());
+
+  public ExtensionFlowProcessingPhase(PolicyManager policyManager) {
+    this.policyManager = policyManager;
+  }
+
+  @Override
+  public boolean supportsTemplate(MessageProcessTemplate messageProcessTemplate) {
+    return messageProcessTemplate instanceof ExtensionFlowProcessingPhaseTemplate;
+  }
+
+  @Override
+  public void runPhase(final ExtensionFlowProcessingPhaseTemplate template, final MessageProcessContext messageProcessContext,
+                       final PhaseResultNotifier phaseResultNotifier) {
+    Work flowExecutionWork = new Work() {
+
+      @Override
+      public void release() {
+
+      }
+
+      @Override
+      public void run() {
+        try {
+          MessageSource messageSource = messageProcessContext.getMessageSource();
```

Comment:
I would prefer to include this commit ASAP since the gateway depends on it to start implementing policies. Perhaps we can create a separate issue to work on once we have defined the need for non-blocking.